Improve logging and handling of errors in s3proxy

MalinAhlberg opened the pull request:
**Related issue(s) and PR(s)**  
This PR closes #413 


**Description**
The original goal was to 
- make sure that errors are logged properly

While working on this, it also made sense to
- make sure errors are properly reported back to the user
- fix erroneous connection to mq in `s3inbox_test`.


**How to test**
Make sure all tests pass. Try making a request with s3cmd with an invalid token, or when either db or mq is disconnected, and make sure errors are logged and reported.

MalinAhlberg commented:
The force-push was to rebase on main, to changes added, so please ignore.

jbygdell commented on sda/cmd/s3inbox/proxy.go:
```diff
@@ -223,6 +228,24 @@ func (p *Proxy) allowedResponse(w http.ResponseWriter, r *http.Request) {
 	_ = s3response.Body.Close()
 }
 
+// Renew the connection to MQ if necessary, then send message
+func (p *Proxy) checkAndSendMessage(jsonMessage []byte, r *http.Request) error {
+	if p.messenger.IsConnClosed() {
+		log.Warning("connection is closed, reconnecting...")
+		m, err := broker.NewMQ(Conf.Broker)
+		p.messenger = m
+
+		return err
+	}
+
+	if err := p.messenger.SendMessage(p.fileIds[r.URL.Path], p.messenger.Conf.Exchange, p.messenger.Conf.RoutingKey, jsonMessage); err != nil {
+
+		return fmt.Errorf("error when sending message to broker: %v", err)
+	}
+
+	return nil
+}
+
```
There is no need to have this as a separate function since it is only used once.
But there are a case missing, when the channel is closed.

```
		if p.messenger.Connection.IsClosed() {
			log.Warning("connection is closed, reconnecting...")
			p.messenger, err = broker.NewMQ(p.messenger.Conf)
			if err != nil {
				log.Errorf("broker error: %v", err)
				reportError(http.StatusInternalServerError, "could not connect to broker", w)

				return
			}
		}
		if p.messenger.Channel.IsClosed() {
			log.Warning("channel is closed, recreating...")
			if err := p.messenger.CreateNewChannel(); err != nil {
				log.Errorf("broker error: %v", err)
				reportError(http.StatusInternalServerError, "could not create channel", w)

				return
			}
		}

		if err := p.messenger.SendMessage(p.fileIds[r.URL.Path], p.messenger.Conf.Exchange, p.messenger.Conf.RoutingKey, jsonMessage); err != nil {
			log.Errorf("sending message failed: %v", err)
			reportError(http.StatusInternalServerError, "communication failure", w)

			return
		}
```

There are also tests missing for the cases where the channel and connection is closed.

MalinAhlberg commented:
New commits: e8a8cab - f3ac792
Fixed the missing cases and added tests, changed http status  for similarity with `api` and `syncapi` and set the correct mq config when reconnecting.

jbygdell commented on sda/cmd/s3inbox/proxy.go:
```diff
@@ -162,35 +163,36 @@ func (p *Proxy) allowedResponse(w http.ResponseWriter, r *http.Request) {
+		err = p.checkAndSendMessage(jsonMessage, r)
+		if err != nil {
+			log.Errorf("broker error: %v", err)
+			reportError(http.StatusServiceUnavailable, "could not connect to broker", w)
```
```suggestion
			p.internalServerError(w, r)
```

jbygdell commented on sda/cmd/s3inbox/proxy.go:
```diff
@@ -223,6 +227,33 @@ func (p *Proxy) allowedResponse(w http.ResponseWriter, r *http.Request) {
 	_ = s3response.Body.Close()
 }
 
+// Renew the connection to MQ if necessary, then send message
+func (p *Proxy) checkAndSendMessage(jsonMessage []byte, r *http.Request) error {
+	if p.messenger.IsConnClosed() {
+		log.Warning("connection is closed, reconnecting...")
+		m, err := broker.NewMQ(p.messenger.Conf)
+		p.messenger = m
+
+		return err
+	}
+
+	if p.messenger.Channel.IsClosed() {
+		log.Warning("channel is closed, reconnecting...")
+		p.messenger.Connection.Close()
+		m, err := broker.NewMQ(p.messenger.Conf)
+		p.messenger = m
+
+		return err
+	}
+
+	if err := p.messenger.SendMessage(p.fileIds[r.URL.Path], p.messenger.Conf.Exchange, p.messenger.Conf.RoutingKey, jsonMessage); err != nil {
+
+		return fmt.Errorf("error when sending message to broker: %v", err)
+	}
+
+	return nil
+}
+
```
```suggestion
// Renew the connection to MQ if necessary, then send message
func (p *Proxy) checkAndSendMessage(jsonMessage []byte, r *http.Request) error {
	var err error
	if p.messenger.IsConnClosed() {
		log.Warning("connection is closed, reconnecting...")
		p.messenger, err = broker.NewMQ(p.messenger.Conf)
		if err != nil {
			return err
		}
	}

	if p.messenger.Channel.IsClosed() {
		log.Warning("channel is closed, recreating...")
		if err := p.messenger.CreateNewChannel(); err != nil {
			return err
		}
	}

	if err := p.messenger.SendMessage(p.fileIds[r.URL.Path], p.messenger.Conf.Exchange, p.messenger.Conf.RoutingKey, jsonMessage); err != nil {
		return fmt.Errorf("error when sending message to broker: %v", err)
	}

	return nil
}

```

jbygdell commented on sda/cmd/s3inbox/proxy_test.go:
```diff
@@ -222,6 +222,60 @@ func (suite *ProxyTests) TestServeHTTPS3Unresponsive() {
+
+	// Test that the correct status code is returned when mq connection can't be created
+	proxy.messenger.Conf.Port = 123456
+	proxy.messenger.Channel.Close()
+	assert.True(suite.T(), proxy.messenger.Channel.IsClosed())
```
```suggestion
	proxy.messenger.Connection.Close()
	assert.True(suite.T(), proxy.messenger.Connection.IsClosed())
```

jbygdell commented on sda/cmd/s3inbox/proxy.go:
```diff
@@ -273,7 +304,7 @@ func (p *Proxy) prependBucketToHostPath(r *http.Request) {
 	str, err := url.ParseRequestURI(r.URL.Path)
 	if err != nil || str.Path == "" {
-		log.Errorf("failed to get path from query (%v)", r.URL.Path)
+		log.Warningf("failed to get path from query (%v)", r.URL.Path)
```
Here we need to return an error because if `str.Path` is `""` the `strings.Split` below will fail with a `nil pointer dereference` error.

MalinAhlberg replied:
good catch! implemented in 7d8f023

jbygdell commented on sda/cmd/s3inbox/proxy.go:
```diff
@@ -162,35 +163,36 @@ func (p *Proxy) allowedResponse(w http.ResponseWriter, r *http.Request) {
+		err = p.database.UpdateFileEventLog(p.fileIds[r.URL.Path], "uploaded", p.fileIds[r.URL.Path], "inbox", "{}", string(jsonMessage))
+		if err != nil {
+			log.Error(err)
+			reportError(http.StatusInternalServerError, "could not connect to db", w)
```
```suggestion
			p.internalServerError(w, r)
```

jbygdell commented:
> @jbygdell or anyone else who has an opinion: Do you prefer 500 over 503 when the db or mq is unavailable? I have no strong opinion, but 503 is used in syncapi and api (eg [here](https://github.com/neicnordic/sensitive-data-archive/blob/c4019cc411eb4fa45d2c5ec0fabe88d4c57b82bb/sda/cmd/syncapi/syncapi.go#L114)) and I think consistency is nice.

The workings are a bit different.
In the syncapi it continuously checks if the connection/channel with the MQ i alive and if it detects that it needs to be recreated it sets the status to 503 to prevent incoming connections from going through until it has had time to reconnect with the MQ.

for the s3inbox we report on an event that a user initiated and there we should just say that something went wrong so the user can try again.

MalinAhlberg commented:
Your comments should be fixed now @jbygdell . I will clean up the git history once you have approved the changes.

jbygdell commented on sda/cmd/s3inbox/proxy.go:
```diff
@@ -143,7 +148,8 @@ func (p *Proxy) allowedResponse(w http.ResponseWriter, r *http.Request) {
 		log.Debugf("fileId: %v", p.fileIds[r.URL.Path])
 		if err != nil {
-			log.Errorf("failed to register file in database: %v", err)
+
```
```suggestion
```

jbygdell commented on sda/cmd/s3inbox/proxy.go:
```diff
@@ -223,6 +227,35 @@ func (p *Proxy) allowedResponse(w http.ResponseWriter, r *http.Request) {
+	if p.messenger.IsConnClosed() {
+		log.Warning("connection is closed, reconnecting...")
+		p.messenger, err = broker.NewMQ(p.messenger.Conf)
+
```
```suggestion
```

jbygdell commented on sda/cmd/s3inbox/proxy.go:
```diff
@@ -223,6 +227,35 @@ func (p *Proxy) allowedResponse(w http.ResponseWriter, r *http.Request) {
+	if p.messenger.Channel.IsClosed() {
+		log.Warning("channel is closed, recreating...")
+		err := p.messenger.CreateNewChannel()
+
```
```suggestion
```

jbygdell commented on sda/cmd/s3inbox/proxy.go:
```diff
@@ -223,6 +227,35 @@ func (p *Proxy) allowedResponse(w http.ResponseWriter, r *http.Request) {
+	}
+
+	if err := p.messenger.SendMessage(p.fileIds[r.URL.Path], p.messenger.Conf.Exchange, p.messenger.Conf.RoutingKey, jsonMessage); err != nil {
+
```
```suggestion
```

jbygdell reviewed (approved):
Some unnecessary empty lines but other than that it looks good.

MalinAhlberg commented:
07f6fe5 is for cases as when the mq container is stopped.

pahatz commented on sda/cmd/s3inbox/proxy_test.go:
```diff
@@ -222,6 +222,60 @@ func (suite *ProxyTests) TestServeHTTPS3Unresponsive() {
+	proxy.messenger.Conf.Port = 123456
+	proxy.messenger.Connection.Close()
+	assert.True(suite.T(), proxy.messenger.Connection.IsClosed())
+	r, _ := http.NewRequest("PUT", "/username/mqunavailbale-file", nil)
```
an insignificant typo here
```suggestion
	r, _ := http.NewRequest("PUT", "/username/mqunavailable-file", nil)
```

dbampalikis reviewed (approved):
LGTM

aaperis reviewed (approved):
Looks good!
